src/socket/communicate.rs

```diff
@@ -0,0 +1,239 @@
+        Message(ReadWriteError),
+    }
+
+    /// Error when initializing connection with the `Listener`.
+    #[derive(Debug)]
+    pub enum InitError {
+        /// `connect()` syscall failed.
+        SocketConnect(std::io::Error),
+        /// Handshake failed (write `ConnectionType`, read `ConnectionAccepted`).
+        ServerHandshake(ReadWriteError),
+    }
+
+    // builder pattern for timeouts?
+
+    impl<R, W> Client<R, W> {
+        /// “Bake” a Client, aka set its communication type (and message type arguments).
+        /// Not exported.
+        fn bake(timeout: Timeout, comm_type: CommunicationType) -> Client<R, W> {
+            Client {
+                comm_type,
+                socket: None,
+                timeout,
+                phantom_r: PhantomData,
+                phantom_w: PhantomData,
+            }
+        }
+
+        /// Connect to the `Listener` listening on `socket_path`.
+        pub fn connect(self, socket_path: &SocketPath) -> Result<Client<R, W>, InitError> {
+            // TODO: check if the file exists and is a socket
+
+            // - connect to `socket_path`
+            let socket = socket_path.connect().map_err(InitError::SocketConnect)?;
+
+            // - send initial message with the CommunicationType
+            // - wait for server to acknowledge connect
+            let _: listener::ConnectionAccepted = ReadWriter::new(&socket)
+                .communicate(self.timeout.clone(), &self.comm_type)
+                .map_err(InitError::ServerHandshake)?;
+
+            Ok(Client {
+                comm_type: self.comm_type,
+                socket: Some(socket),
+                timeout: self.timeout,
+                phantom_r: PhantomData,
+                phantom_w: PhantomData,
+            })
+        }
+
+        /// Read a message returned by the connected `Listener`.
+        pub fn read(self) -> Result<R, Error>
+        where
+            R: serde::de::DeserializeOwned,
+        {
+            let sock = &self.socket.ok_or(Error::NotConnected)?;
+            let rw: ReadWriter<R, W> = ReadWriter::new(sock);
+            rw.read(&self.timeout)
+                .map_err(|e| Error::Message(ReadWriteError::R(e)))
+        }
+
+        /// Write a message to the connected `Listener`.
+        pub fn write(self, mes: &W) -> Result<(), Error>
+        where
+            W: serde::Serialize,
+        {
+            let sock = &self.socket.ok_or(Error::NotConnected)?;
+            let mut rw: ReadWriter<R, W> = ReadWriter::new(sock);
+            rw.write(&self.timeout, mes)
+                .map_err(|e| Error::Message(ReadWriteError::W(e)))
+        }
+    }
+
+    /// Client for the `Ping` communication type.
+    /// Reading and writing messages is bounded by `timeout`.
+    pub fn ping(timeout: Timeout) -> Client<NoMessage, Ping> {
+        Client::bake(timeout, CommunicationType::Ping)
+    }
+
+    // /// Client for the `DaemonListener` communication type
```

Comment:
Let's delete this. (I'd do it, but this is a lot of lines to add single-line suggestions.)